test/common/index.js

```diff
@@ -519,6 +519,48 @@ function mustNotCall(msg) {
   };
 }
 
+const _mustNotMutateProxies = new WeakMap();
+
+function mustNotMutate(original) {
+  if (original === null || typeof original !== 'object') {
```

Comment:
This was discussed a bit earlier: https://github.com/nodejs/node/pull/43196#discussion_r884860952

Additionally, wrapping functions is very problematic for tests because it effectively creates an entity incomparable to original (https://github.com/nodejs/node/issues/43263). Maybe there are more nuances.

Places where we pass functions and there is a risk of mutation seem to be uncommon. If we pass a function somewhere where it shouldn't be passed, we usually want it to not be accidentally called as well. I think, simply replacing them with `Object.freeze(common.mustNotCall())` (or a specific Proxy with specific `apply` trap which will be able to test its arguments) would much more appropriate.

Reply:
Let's throw if `original` is a function then, to make sure no bug can slip through this surprising edge case.

Reply:
I don't think throwing is a good idea here. It will be troublesome whenever passed object happens to have a nested function (at any depth), and it will break transparent usage when we pass lots of `<any>` values in tests.

Added a warning via https://github.com/nodejs/node/pull/43196/commits/fe9279ad32fc9c7ffcd29bb426003c373999d162.

Reply:
I don't think a documentation warning is enough here, I wouldn't blame collaborators to assume that `mustNotMutate` would protect against any kind of mutation, it feels really weird to have functions treated as primitive when they're not. What's the reason for ignoring functions besides that we don't know yet if it's going to be useful? Looking at the earlier commits of this PR it doesn't look like adding support is much complex.

Reply:
It's weird but common in JS to treat functions as non-objects due to their `typeof`. It's common to treat them as non-objects in Node.js core (default options of `validateObject()` and usual checks for `typeof` are not considering fact that Function can have writable properties, too).

Runtime warning is possible, but it won't be seen in false positive tests and it would produce too much noise in other places.

The main reason is that creating an immutable copy, wrapper or proxy of function creates a function incomparable to original. Which means, `assert.deepEqual(target, mustNotMutate(target))` will throw if `target` has nested function or is a function itself.
It will break any deep-equality test for object returned from this function as soon as it gets nested function: in fact, even this one will fail if a function will be added to `original` properties:
https://github.com/nodejs/node/blob/1201eac8e809a9dde552d886ec52cd1c39bccd8d/test/parallel/test-common-must-not-mutate.mjs#L18-L19

The only workaround I see is adding a dirty hack to `assert` methods. Roughly, `mustNotMutate` can add a symbol property on returned proxy, referencing original target, and `assert` will be ok if `a[kCustomEqual] === b || a === b[kCustomEqual]`. But this might introduce more problems that it solves.

Reply:
> It's weird but common in JS to treat functions as non-objects due to their `typeof`

I'm fine with treating them as non-objects, but not treating them as primitives.

What about this:

```js
function mustNotMutate(target){
  if(typeof target === 'function')
    return Object.freeze(target);
  }
  // rest of the function as is here...
  // ...
}
```

That way, the strict equality is maintained, and the test should fail is something's trying the mutate the function object.

Reply:
I'm sceptical about `Object.freeze` because:
1. It changes the original target (and the worst thing is that it cannot be reverted).
2. It's not deep. This either moves edge case one step deeper ("function properties will be constant but still mutable"), or requires mutation on original function by replacing all its properties with proxies.
Freezing recursively is not an option since it would affect all nested Objects as well.

If you have a strong opinion that one of these options is better than current state, it can be done. I just feel that having an exception that allows mutable functions is a worse tradeoff than implicit side effects on target or its deeply nested entities.

Freezing functions might lead to much more obscure results, for example:
```js
const fn = util.promisify(()=>{});
Object.freeze(fn);

util.promisify(fn); // if fn was promisified, it should return the promisified function directly
// Uncaught TypeError: Cannot redefine property: Symbol(nodejs.util.promisify.custom)

util.promisify(util.callbackify(fn));
// Uncaught TypeError: Cannot redefine property: Symbol(nodejs.util.promisify.custom)
```

Reply:
> Freezing functions might lead to much more obscure results

OK, but not freezing would lead to test success, which is arguably even worse.


> 1. It changes the original target (and the worst thing is that it cannot be reverted)

Yes that's quite bad indeed.


Alternative options I can think of:

- rename the function to `canNotMutateNonFunctions` or `canNotMutateObjectDeep` to clarify that functions are not handled.
- add an option `treatFunctionsAsPrimitives`, when `true` return the function unchanged, when `false` returns a Proxy instead.

wdyt?

Reply:
`canNotMutateObjectDeep` sounds great. :+1:

I think, we can keep a warning in docs and update the name for now.
If there will be a need for different behaviour, it can be reimplemented in that specific test.
If a need to adjust default behaviour or add frequently-wanted options will arise later, this function can be changed basing on practical experience and examples.